File: JavaPractice/src/SudokuGame.java
import java.util.*;
/**
 * @author divyesh.shah
 *
 * Determine if a 9 x 9 Sudoku board is valid. Only the filled cells need to be validated according to the following rules:
 *
 * Each row must contain the digits 1-9 without repetition.
 * Each column must contain the digits 1-9 without repetition.
 * Each of the nine 3 x 3 sub-boxes of the grid must contain the digits 1-9 without repetition.
 * Note:
 *
 * A Sudoku board (partially filled) could be valid but is not necessarily solvable.
 * Only the filled cells need to be validated according to the mentioned rules.
 *
 * Input: board =
 * [["5","3",".",".","7",".",".",".","."]
 * ,["6",".",".","1","9","5",".",".","."]
 * ,[".","9","8",".",".",".",".","6","."]
 * ,["8",".",".",".","6",".",".",".","3"]
 * ,["4",".",".","8",".","3",".",".","1"]
 * ,["7",".",".",".","2",".",".",".","6"]
 * ,[".","6",".",".",".",".","2","8","."]
 * ,[".",".",".","4","1","9",".",".","5"]
 * ,[".",".",".",".","8",".",".","7","9"]]
 * Output: true
 *
 * Input: board =
 * [["8","3",".",".","7",".",".",".","."]
 * ,["6",".",".","1","9","5",".",".","."]
 * ,[".","9","8",".",".",".",".","6","."]
 * ,["8",".",".",".","6",".",".",".","3"]
 * ,["4",".",".","8",".","3",".",".","1"]
 * ,["7",".",".",".","2",".",".",".","6"]
 * ,[".","6",".",".",".",".","2","8","."]
 * ,[".",".",".","4","1","9",".",".","5"]
 * ,[".",".",".",".","8",".",".","7","9"]]
 * Output: false
 * Explanation: Same as Example 1, except with the 5 in the top left corner being modified to 8.
 * Since there are two 8's in the top left 3x3 sub-box, it is invalid.
 */
public class SudokuGame {
    public boolean isValidSudoku(char[][] board) {
        // Initialize hash sets for rows, columns and sub-boxes
        Set<Character>[] rows = new HashSet[9];
        Set<Character>[] cols = new HashSet[9];
        Set<Character>[] boxes = new HashSet[9];
        for (int i = 0; i < 9; i++) {
            rows[i] = new HashSet<Character>();
            cols[i] = new HashSet<Character>();
            boxes[i] = new HashSet<Character>();
        }

        // Traverse through the board and check for each cell
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                char c = board[i][j];
                if (c != '.') {
                    // Check for duplicate values in row, column and sub-box
                    if (!rows[i].add(c) || !cols[j].add(c) || !boxes[(i/3)*3+j/3].add(c)) {
                        return false;
                    }
                }
            }
        }
        return true;
    }
    /*
    * [["5","3",".",".","7",".",".",".","."],["6",".",".","1","9","5",".",".","."],[".","9","8",".",".",".",".","6","."],["8",".",".",".","6",".",".",".","3"],["4",".",".","8",".","3",".",".","1"],["7",".",".",".","2",".",".",".","6"],[".","6",".",".",".",".","2","8","."],[".",".",".","4","1","9",".",".","5"],[".",".",".",".","8",".",".","7","9"]]
    *
    * [["8","3",".",".","7",".",".",".","."],["6",".",".","1","9","5",".",".","."],[".","9","8",".",".",".",".","6","."],["8",".",".",".","6",".",".",".","3"],["4",".",".","8",".","3",".",".","1"],["7",".",".",".","2",".",".",".","6"],[".","6",".",".",".",".","2","8","."],[".",".",".","4","1","9",".",".","5"],[".",".",".",".","8",".",".","7","9"]]
    * */

    public void solveSudoku(char[][] board) {
        solveSudokuHelper(board, 0, 0);
    }

    private boolean solveSudokuHelper(char[][] board, int row, int col) {
        if (col == 9) {  // end of row, move to next row
            row++;
            col = 0;
        }
        if (row == 9) {  // finished all rows, puzzle solved
            return true;
        }
        if (board[row][col] != '.') {  // skip non-empty cells
            return solveSudokuHelper(board, row, col+1);
        }
        for (char ch = '1'; ch <= '9'; ch++) {  // try all possible values
            if (isValid(board, row, col, ch)) {
                board[row][col] = ch;
                if (solveSudokuHelper(board, row, col+1)) {
                    return true;
                }
                board[row][col] = '.';  // backtrack
            }
        }
        return false;
    }

    private boolean isValid(char[][] board, int row, int col, char ch) {
        for (int i = 0; i < 9; i++) {  // check row and column
            if (board[row][i] == ch || board[i][col] == ch) {
                return false;
            }
        }
        int subgridRow = (row / 3) * 3;
        int subgridCol = (col / 3) * 3;
        for (int i = subgridRow; i < subgridRow + 3; i++) {  // check 3x3 subgrid
            for (int j = subgridCol; j < subgridCol + 3; j++) {
                if (board[i][j] == ch) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: PythonPractice/src/SudokuGame.py
from typing import List, Optional, Set

class SudokuGame:
    def is_valid_sudoku(self, board: Optional[List[List[str]]]) -> bool:
        if board is None or len(board) != 9 or any(len(row) != 9 for row in board):
            raise ValueError("Pass a valid 9x9 Sudoku board")

        # Initialize sets for rows, columns, and sub-boxes
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boxes = [set() for _ in range(9)]

        # Traverse through the board and check each cell
        for i in range(9):
            for j in range(9):
                c = board[i][j]
                if c != '.':
                    box_index = (i // 3) * 3 + j // 3
                    if (c in rows[i] or c in cols[j] or c in boxes[box_index]):
                        return False
                    rows[i].add(c)
                    cols[j].add(c)
                    boxes[box_index].add(c)
        return True

    def solve_sudoku(self, board: Optional[List[List[str]]]) -> None:
        if board is None or len(board) != 9 or any(len(row) != 9 for row in board):
            raise ValueError("Pass a valid 9x9 Sudoku board")

        self.solve_sudoku_helper(board, 0, 0)

    def solve_sudoku_helper(self, board: List[List[str]], row: int, col: int) -> bool:
        if col == 9:
            row += 1
            col = 0
        if row == 9:
            return True
        if board[row][col] != '.':
            return self.solve_sudoku_helper(board, row, col + 1)

        for ch in map(str, range(1, 10)):
            if self.is_valid(board, row, col, ch):
                board[row][col] = ch
                if self.solve_sudoku_helper(board, row, col + 1):
                    return True
                board[row][col] = '.'  # backtrack

        return False

    def is_valid(self, board: List[List[str]], row: int, col: int, ch: str) -> bool:
        for i in range(9):
            if board[row][i] == ch or board[i][col] == ch:
                return False
        subgrid_row = (row // 3) * 3
        subgrid_col = (col // 3) * 3
        for i in range(subgrid_row, subgrid_row + 3):
            for j in range(subgrid_col, subgrid_col + 3):
                if board[i][j] == ch:
                    return False
        return True

def main():
    board1 = [
        ["5","3",".",".","7",".",".",".","."],
        ["6",".",".","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        [".",".",".",".","8",".",".","7","9"]
    ]

    board2 = [
        ["8","3",".",".","7",".",".",".","."],
        ["6",".",".","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        [".",".",".",".","8",".",".","7","9"]
    ]

    game = SudokuGame()
    print(game.is_valid_sudoku(board1))  # Output: True
    print(game.is_valid_sudoku(board2))  # Output: False
    game.solve_sudoku(board1)
    for row in board1:
        print(row)

if __name__ == "__main__":
    main()
